### ISSUE TITLE
Coder CLI needs a clearer error when it connects to an incompatible server

### ISSUE THREAD

**Issue by spikecurtis** [coder/coder#13478]
If you connect with v2.12 CLI to an earlier Coder Server, with any command that connects to the agent (`coder ssh`, `coder speedtest`, `coder port-forward`, etc.), you'll see

```
2024-06-05 11:31:59.362 [erro]  failed to dial tailnet v2+ API  error="failed to WebSocket dial: expected handshake response status code 101 but got 400"
2024-06-05 11:31:59.572 [debu]  net.wgengine: ping(fd7a:115c:a1e0:405d:acae:5d4e:7d41:a808): no matching peer
2024-06-05 11:31:59.706 [debu]  dialing Coder tailnet v2+ API
2024-06-05 11:31:59.760 [erro]  failed to dial tailnet v2+ API  error="failed to WebSocket dial: expected handshake response status code 101 but got 400"
2024-06-05 11:31:59.872 [debu]  net.wgengine: ping(fd7a:115c:a1e0:405d:acae:5d4e:7d41:a808): no matching peer
2024-06-05 11:32:00.316 [debu]  dialing Coder tailnet v2+ API
2024-06-05 11:32:00.368 [erro]  failed to dial tailnet v2+ API  error="failed to WebSocket dial: expected handshake response status code 101 but got 400"
2024-06-05 11:32:00.512 [debu]  net.wgengine: ping(fd7a:115c:a1e0:405d:acae:5d4e:7d41:a808): no matching peer
2024-06-05 11:32:01.200 [debu]  net.wgengine: ping(fd7a:115c:a1e0:405d:acae:5d4e:7d41:a808): no matching peer
2024-06-05 11:32:01.266 [debu]  dialing Coder tailnet v2+ API
2024-06-05 11:32:01.318 [erro]  failed to dial tailnet v2+ API  error="failed to WebSocket dial: expected handshake response status code 101 but got 400"
```

I'm pretty sure the server is writing a reasonably helpful HTTP response with an error in the body, but here we only log the status code and it's not helpful.

**Comment by Philip-21** [comment]
Hi @spikecurtis  i Would like to work on this

**Comment by spikecurtis** [comment]
@Philip-21 great! 

Our [contributing guide](https://coder.com/docs/CONTRIBUTING) describes how to get your environment set up to build and test.

Please open a PR with your proposed changes when you are ready.

**Comment by Philip-21** [comment]
Hey @spikecurtis to get more context from this, what command did you run before you were able to spot this error ?

**Comment by spikecurtis** [comment]
> Hey @spikecurtis to get more context from this, what command did you run before you were able to spot this error ?

I updated the description with some example commands

**Comment by Philip-21** [comment]
@spikecurtis , here is what i did , 

I ran my coder-server with `v 2.11`  locally and i started my workspace 

Since my coder server was running on `v 2.11` I decided to run commands on the development codebase   to trigger the coder agent that is running, since the development codebase contains the updates for `v.12`

I ran this command  to get the list of my workspace i got this 

              [nix-shell:~/open-source/coder/coder/cmd/coder]$ go run . list
              
              WORKSPACE                 TEMPLATE  STATUS   HEALTHY  LAST BUILT  CURRENT VERSION     OUTDATED  STARTS AT  STOPS AFTER  
               Philip-Coder/Philip-21  docker    Started  false    50m         thirsty_heyrovsky8  false                1d           



I decided to run commands to trigger the agent I got this 
                
            [nix-shell:~/open-source/coder/coder/cmd/coder]$ go run . agent debug-address
            
            2024-06-16 07:05:25.123 [info]  agent is starting now  url=""  auth=token  version=v0.0.0-devel
            2024-06-16 07:05:25.123 [debu]  http server listening  addr=127.0.0.1:6060  name=pprof
            Encountered an error running "coder agent", see "coder agent --help" for more information
            error: CODER_AGENT_TOKEN or CODER_AGENT_TOKEN_FILE must be set for token auth
            exit status 1
            
Though i  have logged in already with a token  on my development server I  still get the error above   

I dont know if i am missing something with what i have done so far.

**Comment by Philip-21** [comment]
These are the server errors i get when i trigger various commands from my development codebase
                error= do:
                           github.com/coder/wgtunnel/tunnelsdk.(*Client).Request
                               /home/runner/go/pkg/mod/github.com/coder/wgtunnel@v0.1.13-0.20231127054351-578bfff9b92a/tunnelsdk/client.go:67
                         - Post "https://pit-1.try.coder.app/api/v2/clients": dial tcp: lookup pit-1.try.coder.app on 127.0.0.53:53: server misbehaving
            2024-06-16 06:24:19.752 [erro]  net.devtunnel.wireguard_device: peer(9i3o…UMXM) - Failed to send handshake initiation: write udp 0.0.0.0:60104: sendmmsg: network is unreachable
            2024-06-16 06:24:25.025 [erro]  net.devtunnel.wireguard_device: peer(9i3o…UMXM) - Failed to send handshake initiation: write udp 0.0.0.0:60104: sendmmsg: network is unreachable
            2024-06-16 06:24:30.045 [erro]  net.devtunnel.wireguard_device: peer(9i3o…UMXM) - Failed to send handshake initiation: write udp 0.0.0.0:60104: sendmmsg: network is unreachable
            2024-06-16 06:30:19.547 [erro]  net.devtunnel.wireguard_device: peer(9i3o…UMXM) - Failed to send handshake initiation: write udp 0.0.0.0:60104: sendmmsg: network is unreachable
            2024-06-16 06:35:42.712 [info]  dbpurge: purged old database entries  duration=8.798294ms
            2024-06-16 06:45:42.724 [info]  dbpurge: purged old database entries  duration=11.02481ms
            2024-06-16 06:55:42.733 [info]  dbpurge: purged old database entries  duration=7.433584ms
            2024-06-16 07:05:45.359 [info]  dbpurge: purged old database entries  duration=6.759789ms
            2024-06-16 07:15:45.368 [info]  dbpurge: purged old database entries  duration=7.891576ms
            2024-06-16 07:25:45.375 [info]  dbpurge: purged old database entries  duration=4.757609ms

**Comment by spikecurtis** [comment]
Hey, sorry, I could have sworn I added this information to the description, but it seems my edits were lost.  After starting a workspace with Coder 2.11, you can run one of the following commands with a 2.12 (or `main` branch) CLI:

 * `coder ssh`
 * `coder speedtest`
 * `coder port-forward`
 
 to connect to the workspace.

**Comment by Philip-21** [comment]
> Hey, sorry, I could have sworn I added this information to the description, but it seems my edits were lost. After starting a workspace with Coder 2.11, you can run one of the following commands with a 2.12 (or `main` branch) CLI:
> 
> * `coder ssh`
> * `coder speedtest`
> * `coder port-forward`
> 
> to connect to the workspace.

Thanks I can work with this ?

**Comment by spikecurtis** [comment]
fixed by #13683
